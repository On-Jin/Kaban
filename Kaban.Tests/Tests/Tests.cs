using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using Kaban.Data;
using Kaban.Models;
using Kaban.Tests.Setup;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;
using Xunit.Abstractions;

namespace Kaban.Tests.Tests;

[Collection(nameof(WebAppCollectionFixture))]
public class Tests : IAsyncLifetime
{
    private readonly ITestOutputHelper _testOutputHelper;
    private readonly HttpClient _httpClientDiscord;
    private readonly HttpClient _httpClientShadow;

    private readonly Func<Task> _resetDatabase;
    private readonly WebAppFactory _factory;

    public Tests(WebAppFactory factory, ITestOutputHelper testOutputHelper)
    {
        _factory = factory;
        _testOutputHelper = testOutputHelper;
        _resetDatabase = factory.ResetDatabaseAsync;

        _httpClientDiscord = factory.CreateClient();
        _httpClientDiscord.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue(TestAuthHandler.AuthenticationScheme);
        _httpClientShadow = factory.CreateClient();

        _httpClientShadow.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue(TestAuthHandler.AuthenticationScheme);
        _httpClientShadow.DefaultRequestHeaders.Add(TestHelper.HeaderUserGuid, TestHelper.GuidShadowAuth);
    }

    public Task InitializeAsync() => Task.CompletedTask;

    public Task DisposeAsync() => _resetDatabase();

    private async Task GenerateKabanBoardToShadowUser(AppDbContext db)
    {
        var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
        var boards = await GenerateHelper.GenerateDefaultKabanBoards();
        user.Boards.AddRange(boards);
        await db.SaveChangesAsync();
    }

    [Fact]
    public async Task DeleteOneBoardTestCascade()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        await GenerateKabanBoardToShadowUser(db);

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            user.Boards.RemoveAt(0);
            await db.SaveChangesAsync();
        }
        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            JsonSerializer.Serialize(user, TestHelper.JsonSerializerOptions).MatchSnapshot();
            Assert.Equal(2, db.Boards.Count());
        }
    }

    [Fact]
    public async Task AddGeneratedKabanBoardsTest()
    {
        await _resetDatabase();
        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        await GenerateKabanBoardToShadowUser(db);

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            JsonSerializer.Serialize(user, TestHelper.JsonSerializerOptions).MatchSnapshot();
        }
    }

    [Fact]
    public async Task Test1()
    {
        await _resetDatabase();

        using var scope = _factory.Services.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            var boards = await GenerateHelper.GenerateDefaultKabanBoards();
            user.Boards.AddRange(boards);
            await db.SaveChangesAsync();
        }

        {
            _testOutputHelper.WriteLine(db.Boards.ToList().Count.ToString());
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            user.Boards.RemoveAt(0);
            await db.SaveChangesAsync();
            _testOutputHelper.WriteLine(db.Boards.ToList().Count.ToString());
        }

        {
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            _testOutputHelper.WriteLine(JsonSerializer.Serialize(user, TestHelper.JsonSerializerOptions));
        }
        {
            _testOutputHelper.WriteLine(db.Users.ToList().Count.ToString());
            var user = (await db.Users.FindAsync(new Guid(TestHelper.GuidShadowAuth)))!;
            db.Users.Remove(user);
            await db.SaveChangesAsync();
            _testOutputHelper.WriteLine(db.Users.ToList().Count.ToString());
        }
    }
}