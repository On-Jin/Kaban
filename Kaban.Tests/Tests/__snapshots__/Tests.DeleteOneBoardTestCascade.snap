{
  "Id": "158ef677-f7a1-424a-bea4-0d6d8ec95717",
  "DiscordId": null,
  "DiscordAvatar": null,
  "DiscordUsername": null,
  "Boards": [
    {
      "Id": 2,
      "Name": "Marketing Plan",
      "Columns": [
        {
          "Id": 4,
          "Order": 0,
          "Name": "Todo",
          "MainTasks": [
            {
              "Id": 18,
              "Order": 0,
              "Title": "Plan Product Hunt launch",
              "Description": "",
              "SubTasks": [
                {
                  "Id": 37,
                  "Order": 0,
                  "Title": "Find hunter",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 38,
                  "Order": 1,
                  "Title": "Gather assets",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 39,
                  "Order": 2,
                  "Title": "Draft product page",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 40,
                  "Order": 3,
                  "Title": "Notify customers",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 41,
                  "Order": 4,
                  "Title": "Notify network",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                },
                {
                  "Id": 42,
                  "Order": 5,
                  "Title": "Launch!",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 18
                }
              ],
              "Column": null,
              "ColumnId": 4
            },
            {
              "Id": 19,
              "Order": 1,
              "Title": "Share on Show HN",
              "Description": "",
              "SubTasks": [
                {
                  "Id": 43,
                  "Order": 0,
                  "Title": "Draft out HN post",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 19
                },
                {
                  "Id": 44,
                  "Order": 1,
                  "Title": "Get feedback and refine",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 19
                },
                {
                  "Id": 45,
                  "Order": 2,
                  "Title": "Publish post",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 19
                }
              ],
              "Column": null,
              "ColumnId": 4
            },
            {
              "Id": 20,
              "Order": 2,
              "Title": "Write launch article to publish on multiple channels",
              "Description": "",
              "SubTasks": [
                {
                  "Id": 46,
                  "Order": 0,
                  "Title": "Write article",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                },
                {
                  "Id": 47,
                  "Order": 1,
                  "Title": "Publish on LinkedIn",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                },
                {
                  "Id": 48,
                  "Order": 2,
                  "Title": "Publish on Inndie Hackers",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                },
                {
                  "Id": 49,
                  "Order": 3,
                  "Title": "Publish on Medium",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 20
                }
              ],
              "Column": null,
              "ColumnId": 4
            }
          ],
          "Board": null,
          "BoardId": 2
        },
        {
          "Id": 5,
          "Order": 1,
          "Name": "Doing",
          "MainTasks": [],
          "Board": null,
          "BoardId": 2
        },
        {
          "Id": 6,
          "Order": 2,
          "Name": "Done",
          "MainTasks": [],
          "Board": null,
          "BoardId": 2
        }
      ]
    },
    {
      "Id": 3,
      "Name": "Roadmap",
      "Columns": [
        {
          "Id": 7,
          "Order": 0,
          "Name": "Now",
          "MainTasks": [
            {
              "Id": 21,
              "Order": 0,
              "Title": "Launch version one",
              "Description": "",
              "SubTasks": [
                {
                  "Id": 50,
                  "Order": 0,
                  "Title": "Launch privately to our waitlist",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 21
                },
                {
                  "Id": 51,
                  "Order": 1,
                  "Title": "Launch publicly on PH, HN, etc.",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 21
                }
              ],
              "Column": null,
              "ColumnId": 7
            },
            {
              "Id": 22,
              "Order": 1,
              "Title": "Review early feedback and plan next steps for roadmap",
              "Description": "Beyond the initial launch, we\u0027re keeping the initial roadmap completely empty. This meeting will help us plan out our next steps based on actual customer feedback.",
              "SubTasks": [
                {
                  "Id": 52,
                  "Order": 0,
                  "Title": "Interview 10 customers",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 22
                },
                {
                  "Id": 53,
                  "Order": 1,
                  "Title": "Review common customer pain points and suggestions",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 22
                },
                {
                  "Id": 54,
                  "Order": 2,
                  "Title": "Outline next steps for our roadmap",
                  "IsCompleted": false,
                  "MainTask": null,
                  "MainTaskId": 22
                }
              ],
              "Column": null,
              "ColumnId": 7
            }
          ],
          "Board": null,
          "BoardId": 3
        },
        {
          "Id": 8,
          "Order": 1,
          "Name": "Next",
          "MainTasks": [],
          "Board": null,
          "BoardId": 3
        },
        {
          "Id": 9,
          "Order": 2,
          "Name": "Later",
          "MainTasks": [],
          "Board": null,
          "BoardId": 3
        }
      ]
    }
  ]
}
